File: python/Longest_Palindromic_Substring.py
'''
Created on 2014.8.15

Leetcode : Longest_Palindromic_Substring

Given a string S, find the longest palindromic substring in S. You may assume 

that the maximum length of S is 1000, and there exists one unique longest palindromic substring

'''

class Solution:
    # @return a string
    def longestPalindrome(self, s):
        if not s:
            return s
        nS = '#'
        for char in s:
            nS += char + "#"
        nS += '#'
        maxRadius = 0
        maxCentor = 0
        rightRange = 2;
        id = 1
        arr = [ 1 for i in range(len(nS)) ]
        
        for i in range(1,len(nS)):
            if rightRange > i:
                arr[i] = min( arr[2*id - i] , rightRange - i)
            else:
                arr[i] = 1
            
            
            while i - arr[i] >= 0 and i + arr[i] < len(nS) and nS[i+arr[i]] == nS[i - arr[i]]:
                arr[i] += 1
            if i+arr[i] > rightRange:
                rightRange = i + arr[i]
                id = i
            if arr[i] > maxRadius:
                maxRadius = arr[i]
                maxCentor = i
        
        result = [ char for char in nS[maxCentor - maxRadius + 1 : maxCentor + maxRadius - 1] if char != '#' ]
        return ''.join(result)
            

test = Solution()
s = 'babdfdac'
print test.longestPalindrome(s)

File: Cplusplus/Longest_Palindromic_Substring.cpp
/*
Leetcode : Longest_Palindromic_Substring.cpp
Given a string S, find the longest palindromic substring in S.

You may assume that the maximum length of S is 1000, and there

exists one unique longest palindromic substring.


Manacher's ALGORITHM

*/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;


class Solution {
public:
    string longestPalindrome(string s) {
        string newS = "#";
        for( int i = 0 ; i < s.size() ; i++)
        {
            newS.push_back(s[i]);
            newS.push_back('#');

        }

        vector<int> p(newS.size() , 1);
        int id = 0;
        int maxId = 0;
        int maxV = 0;
        int rightE = 0;
        for(int i = 0 ; i < newS.size() ; i++)
        {
            if (rightE > i)
            {
                p[i] = min(rightE - i , p[2*id - i]);
            }
            while(i+p[i]<newS.size() && i-p[i]>=0 && newS[i+p[i]] == newS[i-p[i]])
                p[i]++;
            if(i+p[i]>rightE)
            {
                rightE = i+p[i];
                id = i;

            }
            if(p[i]>maxV)
            {
                maxId = i;
                maxV = p[i];
            }

        }
        string result = "";
        for(int i = maxId - maxV + 1 ; i < maxId+maxV ; i++ )
        {
            if (newS[i]!='#')
                result.push_back(newS[i]);
        }
        return result;

    }
};

int main()
{
    Solution test = Solution();
    string s = "daomdukomcayjwgmmetypncdeixarhbectjcwftjjtwjrctixtrse\
    hegwlfotpljeeqhntacfihecdjysgfmxxbjfeskvvfqdoedfxriujnoeteleftsjgd\
    sagqvcgcdjbxgmguunhbjxyajutohgbdwqtjghdftpvidkbftqbpeahxbfyamonazvu\
    bzirhqseetaneptnpjbtrtitttxsyjckjvwtrmuwgidkofvobrwrffzrpnxbectsydbv\
    swstfiqranjzhsebjnmprjoirqkgttahrivkcjtitdcpohwwerwgrdivqbltfnigavydx\
    pmitttjjzyrmpaptkczzgnsovebvxezkkovgqegctxacvjfqwtiycnhartzczcgosiquhm\
    dbljjzyrnmffcwnkyzytnsvyzayrieqyrfpxintbbiyrawxlguzaisedwabpzvevlejadzt\
    uclcpwvonehkhknicdksdcnjfaoeaqhcdkdtywilwewadcnaprcasccdcnvdgjdqirrsqwzh\
    qqorlhbgpelpupdvuynzybcqkffnnpocidkkigimsa";
    cout<<test.longestPalindrome(s)<<endl;
    return 0;
}
